sweep/fee_function.go

```diff
@@ -153,18 +156,18 @@ func NewLinearFeeFunction(maxFeeRate chainfee.SatPerKWeight,
 
 	// The starting and ending fee rates are in sat/kw, so we need to
 	// convert them to msat/kw by multiplying by 1000.
-	delta := btcutil.Amount(end - start).MulF64(1000 / float64(confTarget))
+	delta := btcutil.Amount(end - start).MulF64(1000 / float64(l.width))
 	l.deltaFeeRate = mSatPerKWeight(delta)
 
 	// We only allow the delta to be zero if the width is one - when the
 	// delta is zero, it means the starting and ending fee rates are the
 	// same, which means there's nothing to increase, so any width greater
 	// than 1 doesn't provide any utility. This could happen when the
-	// sweeper is offered to sweep an input that has passed its deadline.
+	// budget is too small.
 	if l.deltaFeeRate == 0 && l.width != 1 {
 		log.Errorf("Failed to init fee function: startingFeeRate=%v, "+
 			"endingFeeRate=%v, width=%v, delta=%v", start, end,
-			confTarget, l.deltaFeeRate)
+			l.width, l.deltaFeeRate)
```

Comment:
What is the reason for this check?  Can we remove it?

It seems like the check prevents broadcasting the transaction in cases where the budget is actually enough.  For example:

- `confTarget` is 2
- starting feerate calculated from `confTarget` is 30 sat/vB
- ending feerate for budget is 30 sat/vB

If we broadcast immediately at 30 sat/vB, the transaction is likely to confirm in time.  But currently an error is returned instead, and broadcasting is delayed until the next block when `confTarget` is 1 and the branch above is taken instead.  But we are likely to miss the deadline in that case, since we broadcasted late.

Reply:
As shown in the unit test, when conf target is 2, the max fee rate will be used. The check is added so we don't have a fee rate delta of 0, since in that case the fee func cannot do anything, also indicates the budget is too small.

Reply:
I think it makes sense to keep this case in place and not broadcast the sweep when we still have not reached the deadline (conftarget=2) so that more inputs can eventually increase the endFeeRate. Eventually the sweep is triggered in case the deadline is reached.

Reply:
> In this case, shouldn't we just broadcast with max budget immediately and hope for the best?

I guess the question is where is the hope coming from:)

I don't know if the age of a tx is still a factor for the miners these days. If not and, if the tx gets confirmed at some blocks later, it means the network fee rate goes down -> the fee func will be created then?

The initial intention is to return an error here, so the user can increase the budget - for instance, when the htlc is large and the budget ratio is small, the user can intervene by specifying a larger budget via `BumpFee`.

> It is very unlikely that future inputs will have the same deadline as the current inputs, so they are unlikely to be batched together.

Yeah this is an issue IMO - if the deadlines are already large, I think they can be grouped - like, how different in terms of urgency they have between deadlines of 1008 and 1007 - think there should be a ladder structure, sth like the fee rate bucket we used to have.

Reply:
> I don't know if the age of a tx is still a factor for the miners these days. If not and, if the tx gets confirmed at some blocks later, it means the network fee rate goes down -> the fee func will be created then?

Maybe the fee func will be created then, or maybe not.  How sensitive is bitcoind's fee estimator to temporary drops in fee rate?  If not too sensitive, then the fee func will continue to fail.

And bitcoind's fee estimator is *backwards-looking*, so it will always be delayed in adjusting downward.  So by the time we actually broadcast the transaction, it might be too late (because deadline is missed or because we missed the lower-fee window and mempool fees went back up).

Overall, I think we're better off to have the transaction in the mempool so it can potentially be mined.  Until we broadcast it, it can *never* be mined.

> The initial intention is to return an error here, so the user can increase the budget - for instance, when the htlc is large and the budget ratio is small, the user can intervene by specifying a larger budget via `BumpFee`.

Perhaps an early sanity check specifically for `BumpFee` is more appropriate?  If budget is less than bitcoind's estimate, we can quickly return an error from `BumpFee` instead of processing further.

For contractcourt sweeps, I think it's better to broadcast at max budget than to give up completely.

> Yeah this is an issue IMO - if the deadlines are already large, I think they can be grouped - like, how different in terms of urgency they have between deadlines of 1008 and 1007 - think there should be a ladder structure, sth like the fee rate bucket we used to have.

+1